### DISCUSSION TITLE
Client metrics collections like in System Manager

### DISCUSSION THREAD

**Discussion by wally007** [NetApp/harvest#1059] [Ideas]
Hello,

i noticed that ONTAP and System Manager keep track of all clients  connections.
![image](https://user-images.githubusercontent.com/627956/170860697-64034043-7a65-48f5-9fb1-e85ff949bb11.png)

**Is it possible to extend Harvest to gather statistics per client ?**
(preferably all protocols NFS/CIFS/S3 etc but i dont know if ONTAP keeps track of all protocols, I know it does for NFS clients)

If you'd rather not open this up by default in Harvest, can you please advise how this can be done? 
I've looked at supplied yaml files and i've got no idea where to even begin...

Thank you very much in advance.

**Comment by rahulguptajss** [comment]
@wally007 

**NFS**
I haven't found any Zapi but similiar information as of system manager is available through Rest (From ONTAP 9.7 onwards) [here](https://library.netapp.com/ecmdocs/ECMLP2862544/html/index.html#/NAS/nfs_clients_get) . If you want to try this out then similiar steps needs to be followed for the same as mentioned [here](https://github.com/NetApp/harvest/discussions/1057)

**CIFS**
There is a Zapi cifs-connection-get-iter for which a harvest template can be written. You can explore this Zapi with below command
```bin/zapi --poller POLLERNAME show attrs --api cifs-connection-get-iter```

**S3**
Are any of [these](https://library.netapp.com/ecmdocs/ECMLP2862544/html/index.html#/object-store/all_svm_buckets_collection_get) rest end points matches your requirement? 

Kindly let us know if you face any issues while creating zapi/rest template for these.

**Reply by wally007** [reply]
thanks for the response.

We are interested (for now) only tracking NFS clients activity - but from the link posted I can only infer that there is no way to track client usage statistics such as bandwidth. Is that correct ?

**Reply by rahulguptajss** [reply]
@wally007 Yes, Both CIFS/NFS client statistics can be achieved through ZapiPerf Collector in Harvest. You would need to enable client statistics first on the cluster. Steps available [here](https://kb.netapp.com/Advice_and_Troubleshooting/Data_Storage_Software/ONTAP_OS/How_to_view_per_client_statistics_in_ONTAP_9_and_later)

I tried below commands in ONTAP.

```
set -privilege advanced
statistics start -object client
```

Now you can explore available counters for ```client``` object as below in Harvest.

```bin/zapi -p POLLERNAME show counters --object client```

Based on these counters, you can create a new harvest template for this object. Does these counters help?

**Reply by wally007** [reply]
thank you for that. 
So far i only looked at the ONTAP documentation regarding enabling -**object client** but i didnt find what kind of penalty does it introduce if one leaves it on all the time.

Is it safe to leave it enabled 24/7/365 ?

The idea is that we want to monitor how much bandwidth each client consumes - which in our case is about 20-30 NFS clients (ESXi hosts mounting NFS shares and some Linux VM's) 
I'm a bit of hesitant since when I enabled the service, it provided me with the "sample-id" (which I inferred means its not meant to stay/remain enabled service for 24/7, but maybe i'm wrong ?)

(fyi, our clusters are all AFF's and we never go above 50% of CPU usage)

**Reply by cgrinds** [reply]
@wally007 we had the same question about leaving enabled 24/7. I suspect with the small number of clients, it will be fine, but we'll send some emails and get an answer.

**Comment by rahulguptajss** [comment]
@wally007 Command earlier shared **will not help** as it only collects data from the command start time. As the time grows larger, stats will be less useful.

```
set -privilege advanced
statistics start -object client
```

There is another way to collect client stats. Below are the steps.

1. Enable client stats as mentioned below.

<img width="898" alt="image" src="https://user-images.githubusercontent.com/25551691/173315184-4fe755ca-1c52-4576-88fd-4bd0fbfa4948.png">

2. Create a ZapiPerf template for object `client` . Below is the harvest command to explore counters available.

`bin/zapi -p POLLERNAME show counters --object client`

**Note:** As mentioned in Step 1 screenshot, Enabling client-stats may impact system performance **significantly**. We don't have any data to measure the impact. Possibly depends on number of clients system is monitoring. Given the warning from Ontap, It is probably not safe to leave it open for long time.

**Comment by wally007** [comment]
thanks for the follow up. 
I guess this is a no-go for production then - even with limited amount of clients (up to 50)

**Reply by rahulguptajss** [reply]
Yes, It's better to try it in pre-prod to get some data. If possible, I would suggest to open an Ontap support ticket to ask the impact on performance for your environment.
